server/grpc/server.go

```diff
@@ -0,0 +1,151 @@
+		}, nil
+	}
+	return nil, fmt.Errorf("method not implemented: %s", in.Method)
+}
+
+// START TOPIC SUB
+
+// AddTopicEventHandler adds provided topic to the list of server subscriptions
+func (s *Server) AddTopicEventHandler(topic string, fn func(ctx context.Context, event *event.TopicEvent) error) {
+	s.topicSubscriptions[topic] = fn
+}
+
+// ListTopicSubscriptions is called by Dapr to get the list of topics the app wants to subscribe to. In this example, we are telling Dapr
+// To subscribe to a topic named TopicA
+func (s *Server) ListTopicSubscriptions(ctx context.Context, in *empty.Empty) (*pb.ListTopicSubscriptionsResponse, error) {
+	subs := make([]*pb.TopicSubscription, 0)
+	for k := range s.topicSubscriptions {
+		sub := &pb.TopicSubscription{
+			Topic: k,
+		}
+		subs = append(subs, sub)
+	}
+
+	return &pb.ListTopicSubscriptionsResponse{
+		Subscriptions: subs,
+	}, nil
+}
+
+// OnTopicEvent fired whenever a message has been published to a topic that has been subscribed. Dapr sends published messages in a CloudEvents 0.3 envelope.
+func (s *Server) OnTopicEvent(ctx context.Context, in *pb.TopicEventRequest) (*empty.Empty, error) {
+	if val, ok := s.topicSubscriptions[in.Topic]; ok {
+		e := &event.TopicEvent{
+			Topic:           in.Topic,
+			Data:            in.Data,
+			DataContentType: in.DataContentType,
+			ID:              in.Id,
+			Source:          in.Source,
+			SpecVersion:     in.SpecVersion,
+			Type:            in.Type,
+		}
+		err := val(ctx, e)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error handling topic event: %s", in.Topic)
+		}
+	}
+	return &empty.Empty{}, nil
+}
+
+// START BINDING
+
+// AddBindingEventHandler add the provided handler to the server binding halder collection
+func (s *Server) AddBindingEventHandler(name string, fn func(ctx context.Context, in *event.BindingEvent) error) {
+	s.bindingHandlers[name] = fn
+}
+
+// ListInputBindings is called by Dapr to get the list of bindings the app will get invoked by. In this example, we are telling Dapr
+// To invoke our app with a binding named storage
+func (s *Server) ListInputBindings(ctx context.Context, in *empty.Empty) (*pb.ListInputBindingsResponse, error) {
+	list := make([]string, 0)
+	for k := range s.bindingHandlers {
+		list = append(list, k)
+	}
+
+	return &pb.ListInputBindingsResponse{
+		Bindings: list,
+	}, nil
+}
+
+// OnBindingEvent gets invoked every time a new event is fired from a registered binding. The message carries the binding name, a payload and optional metadata
+func (s *Server) OnBindingEvent(ctx context.Context, in *pb.BindingEventRequest) (*pb.BindingEventResponse, error) {
+	if val, ok := s.bindingHandlers[in.Name]; ok {
+		e := &event.BindingEvent{
+			Name:     in.Name,
+			Data:     in.Data,
+			Metadata: in.Metadata,
+		}
+		err := val(ctx, e)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error executing %s binding", in.Name)
+		}
+		return &pb.BindingEventResponse{}, nil
```

Comment:
As you may know, BindingEventResponse can store value to statestore and send the event to multiple output binding. Do you have the plan to support it? 

On the other hand, I do not understand why we have this feature in dapr runtime. It will make user confused because they can directly call dapr api explictly. need to discuss this functionality.

Reply:
Not sure I follow, how would that impact the user serving code?

Reply:
in short, have you considered to use the properties in BindingEventResponse{} ? 

https://github.com/dapr/dapr/blob/841849a6a15b27a1dd3877b332984d1a8ddfbbdd/dapr/proto/runtime/v1/appcallback.proto#L87-L109

Spec - https://github.com/dapr/docs/blob/master/reference/api/bindings_api.md#http-response-body-optional

Reply:
We can look into implementing something there in the future, given the required understanding of gRPC from the user, I'm wondering if that user would be even using a "convenience layer" to access Dapr. They can use the Dapr gRPC API itself

Reply:
@youngbupark it's good to bring it up.

@mchmarny we should open an issue to track this.

I don't see this being a blocker for this PR.

Reply:
closing since this is outdated